File: sdks/nodejs/src/retry.ts
/**
 * Retry logic with exponential backoff
 */

import { TollMeshError, ErrorCode } from './errors';

export interface RetryConfig {
  maxRetries?: number;
  baseDelay?: number;
  maxDelay?: number;
  jitter?: boolean;
  backoffMultiplier?: number;
}

/**
 * Calculate delay for retry attempt with exponential backoff
 */
export function calculateDelay(
  attempt: number,
  config: RetryConfig = {}
): number {
  const {
    baseDelay = 1000,
    maxDelay = 60000,
    jitter = true,
    backoffMultiplier = 2.0,
  } = config;

  let delay = Math.min(
    baseDelay * Math.pow(backoffMultiplier, attempt),
    maxDelay
  );

  if (jitter) {
    const jitterAmount = delay * 0.25;
    delay += Math.random() * jitterAmount * 2 - jitterAmount;
  }

  return Math.max(0, Math.floor(delay));
}

/**
 * Check if error is retryable
 */
export function isRetryable(error: any): boolean {
  if (error instanceof TollMeshError) {
    return error.isRetryable();
  }
  // Retry on network errors, timeouts
  return error.code === 'ECONNREFUSED' ||
    error.code === 'ENOTFOUND' ||
    error.code === 'ETIMEDOUT' ||
    error.message?.includes('timeout');
}

/**
 * Sleep for specified milliseconds
 */
export function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Retry a function with exponential backoff
 */
export async function withRetry<T>(
  fn: () => Promise<T>,
  config: RetryConfig = {}
): Promise<T> {
  const maxRetries = config.maxRetries ?? 3;
  let lastError: any;

  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      return await fn();
    } catch (error) {
      lastError = error;

      if (!isRetryable(error)) {
        throw error;
      }

      if (attempt < maxRetries) {
        const delay = calculateDelay(attempt, config);
        await sleep(delay);
      }
    }
  }

  throw lastError;
}

/**
 * Decorator for retrying async functions
 */
export function retry(config: RetryConfig = {}) {
  return function <T extends any[], R>(
    target: any,
    propertyKey: string,
    descriptor: PropertyDescriptor
  ): PropertyDescriptor {
    const originalMethod = descriptor.value;

    descriptor.value = async function (...args: T) {
      return withRetry(
        () => originalMethod.apply(this, args),
        config
      );
    };

    return descriptor;
  };
}

File: sdks/rust/src/retry.rs
use crate::TollMeshError;
use std::future::Future;
use std::time::Duration;
use tokio::time::sleep;

pub struct RetryConfig {
    pub max_retries: u32,
    pub base_delay: Duration,
    pub max_delay: Duration,
    pub jitter: bool,
    pub backoff_multiplier: f64,
}

impl Default for RetryConfig {
    fn default() -> Self {
        Self {
            max_retries: 3,
            base_delay: Duration::from_millis(1000),
            max_delay: Duration::from_secs(60),
            jitter: true,
            backoff_multiplier: 2.0,
        }
    }
}

impl RetryConfig {
    pub fn calculate_delay(&self, attempt: u32) -> Duration {
        let base = self.base_delay.as_millis() as f64;
        let mut delay = (base * self.backoff_multiplier.powi(attempt as i32))
            .min(self.max_delay.as_millis() as f64) as u64;

        if self.jitter {
            let jitter_amount = delay / 4;
            delay = (delay - jitter_amount + rand::random::<u64>() % (jitter_amount * 2))
                .max(0);
        }

        Duration::from_millis(delay)
    }
}

pub async fn with_retry<F, Fut, T>(
    mut f: F,
    config: &RetryConfig,
) -> Result<T, TollMeshError>
where
    F: FnMut() -> Fut,
    Fut: Future<Output = Result<T, TollMeshError>>,
{
    let mut last_error = None;

    for attempt in 0..=config.max_retries {
        match f().await {
            Ok(result) => return Ok(result),
            Err(err) => {
                if !err.is_retryable() {
                    return Err(err);
                }

                last_error = Some(err);

                if attempt < config.max_retries {
                    let delay = config.calculate_delay(attempt);
                    sleep(delay).await;
                }
            }
        }
    }

    Err(last_error.unwrap_or_else(|| {
        TollMeshError::new(crate::ErrorCode::Internal, "Retry failed")
    }))
}
